students/Pirouz_N/lesson04/mailroom.py

```diff
@@ -0,0 +1,185 @@
+    with open('data/donor_db.txt', 'w') as output_file:
+        for db_key in donor_db:
+            output_file.write(('{}: ' + '{}, ' * (len(donor_db[db_key]) - 1) + '{}\n')
+                              .format(db_key, *donor_db[db_key]))
+            # Or using f string
+            # output_file.write(f'{db_key}: {str(donor_db[db_key])[1:len(str(donor_db[db_key])) - 1]}\n')
+
+
+def populate_db_from_file():
+    """"This function reads donors from database in donor_db.txt located at current working sub directory data."""
+    try:
+        with open('data/donor_db.txt', 'r') as input_file:
+            for line in input_file:
+                colon_sep_list = line.split(':')
+                donor_db[colon_sep_list[0]] = [float(item.strip()) for item in colon_sep_list[1].strip().split(',')]
+    except TypeError:
+        print('Database data has been corrupted. Please restore a previously functioning backup.')
+    except FileNotFoundError:
+        print('Database file is no available in the current working directory folder.')
+
+
+def exit_loop():
+    """"This function allows the user to exit main menu using global condition."""
+    global continue_loop
+    continue_loop = False
+
+
+def main_menu():
+    """"This function makes the menu portion of the program work."""
+    global continue_loop
+    continue_loop = True
+    option_dict = {'1': send_thank_you_after_donation, '2': create_report, '3': send_thank_you_to_all, '4': exit_loop}
+    while continue_loop:
+        option = input('Please select of the three options: Send a Thank You, Create a Report, send letters to all '
+                       'donors, or Quit.\n'
+                       'Type 1 and hit enter if you wish to Send a Thank You.\n'
+                       'Type 2 and hit enter if you wish to Create a Report.\n'
+                       'Type 3 and hit enter if you wish to send letters to all donors.\n'
+                       'Type 4 and hit enter if you wish to Quit.\n'
+                       'Please make your selection:\n')
+        try:
+            option_dict[option]()
+        except KeyError:
+            print('Please only type 1 or 2 or 3 or 4 and hit enter. Any other option will not be acceptable.')
+
+
+def create_report():
+    """"This function will create a report."""
+    # Printing table
+    print('{:<30.30s}\t|{:^16.16s}\t|{:^12.12s}\t|{:^16.16s}'
+          .format('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift'))
+    print('------------------------------------------------------------------------------------------------')
+
+    # printing donor table
+    for donor in donor_db:
+        # printing donor table
+        print(
+            '{:<30.30s}\t ${:>15.15s}\t{:>12.12s}\t ${:>15.15s}'
+            .format(donor, format(float(sum(donor_db[donor])), '.2f'), str(len(donor_db[donor])),
+                    format(float(sum(donor_db[donor]) / len(donor_db[donor])), '.2f')))
+    print('\n')
```

Comment:
missing sorting.

Reply:
Totally right. Just corrected it!

Best Regards
Pirouz Naghavi